design/legacy-design/device-service/discovery.md

```diff
@@ -0,0 +1,79 @@
+## Dynamic Device Discovery
+
+### Overview
+
+Some device protocols allow for devices to be discovered automatically. A Device
+Service may include a capability for discovering devices and creating the
+corresponding Device objects within EdgeX. A framework for doing so will be
+implemented in the Device Service SDKs.
+
+The discovery process will operate as follows:
+
+* Discovery is triggered either on an internal timer or by a call to a REST endpoint
+* The SDK will call a function provided by the DS implementation to request a
+device scan
+* The implementation calls back to the SDK with details of devices which it has found
+* The SDK filters these devices against a set of acceptance criteria
+* The SDK adds accepted devices in core-metadata. These are now available in
+the EdgeX system
+
+### Triggering Discovery
+
+A boolean configuration value `Device/Discovery` defaults to false. If this
+value is set true, and the DS implementation supports discovery, discovery is
+enabled.
+
+The SDK will respond to POST requests on the the /discovery endpoint. This call
+will return one of the following codes:
+
+* 200: discovery has been triggered or is already running
+* 423: the service is locked (admin state) or disabled (operating state)
+* 500: unknown or unanticipated issues exist
+* 501: discovery is not supported in this implementation
+* 503: discovery is disabled by configuration
+
+In the case where discovery is triggered, the discovery process will run in a
+new thread or goroutine, so that the REST call may return immediately.
+
+An integer configuration value `Device/DiscoveryInterval` defaults to zero. If
+this value is set to a positive value, and discovery is enabled, the discovery
+process will be triggered at the specified interval (in seconds).
+
+### Finding Devices
+
+When discovery is triggered, the SDK calls the implementation function provided
+by the Device Service. This should perform whatever protocol-specific procedure
+is necessary to find devices, and for each device found, call the filtered
+device addition function provided by the SDK.
+
+The information required for a found device is as follows:
+
+* An autogenerated device name
+* The Protocol Properties of the device
+* Optionally, a description string
+* Optionally, a list of label strings
+
+**Rationale:** *An alternative design would have the implementation function return
+a collection of discovered devices to the SDK. Using a callback mechanism instead has the following advantages:*
+
+* *Avoids collection processing in languages without builtin support for such, eg C*
+* *Allows for asynchronous operation, where the DS implementation initiates discovery and returns immediately, then when results come back (eg, from a network), handler functions in the implementation call the filtered device addition function themselves*
```

Comment:
Please reword for clarity based on discussion Device Services WG (13-Jan-2020)

Reply:
The one thing I'd like to see clarified (as was discussed at the WG meeting that Trevor references) is whether the results of discovery are passed back to the SDK via a single callback, or whether each new device discovered results in a callback. My preference would be for the former (all results are passed back as a single callback).

We also discussed whether the returned list if filtered by the DS impl or whether it always blindly returns *all* devices it finds regardless of whether they're already known to the DS.

Reply:
Reworded. Also specified that the operation may be batched, to take advantage of future API developments.